Make cross-compiling with the crossSystem argument work

DavidEGrayson opened the pull request:
###### Motivation for this change

I would like to be able to compile software for systems that do not have Nix.  It looks like nixpkgs has a `crossSystem` argument that should allow this, but when I tried to use it, I encountered a bunch of problems.

The main problem was that the `crossSystem` argument affected a bunch of native derivations, such as gcc5, so when I tried to cross-compile, Nix would needlessly try to rebuild the stdenv, and encounter some error.

Another issue was that the cross compilers were built using the bootstrap tools instead of the normal stdenv, and they were built using gcc.cc.override which seems invalid to me.  Fixing this issue was the final thing that allowed me to successfully cross-compile for the Raspberry Pi.

See my detailed commit messages for more info.  (I actually redid the commit history before submitting this pull request to make it easy for reviewers to follow.)  Thanks!
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [ ] NixOS
  - [ ] OS X
  - [x] Linux
- [x] Tried to run `nix-shell -p nox --run "nox-review wip --against d4eac027"` but its [error message](https://gist.github.com/DavidEGrayson/efc54d4637c7e521227abb53eb1b3030) did not make much sense to me.  It looks like it tried to evaluate `libcCross1` without specifying a `crossSystem` argument, which is invalid (just like it would also be invalid to evaluate libcCross).
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

The main way I tested this PR was by applying `nix-build` to [this Nix expression](https://gist.github.com/DavidEGrayson/8d61032e523b3bc8bbca0c4b281afb52).  That Nix expression does several things: it ensures that the gcc5 in my pull request does not get affected by the crossSystem argument.  It also ensures that my pull request does not affect any native GCC derivations, so we will not need a mass rebuild.  It also compiles the "hello" package for the Raspberry Pi.

---
###### Future work
- I am only touching the GCC 5 recipe in this pull request, but we should apply my fixes to GCC 6 after they are accepted.
- The test I added in the new directory `test` should be automatically run by Hydra.  I do not know how to set that up.
- Note that this pull request only makes cross-compiling work if the build system is Linux.  A bit more work would be needed to make cross-compiling work on Darwin.

DavidEGrayson committed:
top-level: Cleaned up top-level/stdenv.nix and the way it is called.

No more mutual importing between top-level/default.nix and
top-level/stdenv.nix.  Instead of having stdenv.nix import
default.nix, we will have default.nix construct the allPackages
function and pass it as an argument to stdenv.nix.

No more use of misleading names "self" and "super" in stdenv.nix.
Those names have a specific meaning given to them by lib.extend, which
did not apply.

Incorporate all topLevelArguments into allPackages.  Previously,
allPackages used default values for bootStdenv, noSysDirs,
crossSystem, and platform, ignoring the values passed by the user to
default.nix.  That behavior seemed surprising and likely to cause
bugs, so I changed it.

In stdenv.nix, use better indentation to make it clear that there are
four basic ways to choose a stdenv.

DavidEGrayson committed:
Removed a hack in pkgs/top-level/default.nix in that caused lots of
native derivations (e.g. gcc) to change when crossSystem is specified.

A common sense principle of the crossSystem argument is that its value
should not affect the native derivations.

Instead of using that hack, we use a smarter stdenvOverrides system
that allows us to just change the nativeDrvs while leaving the
crossDrvs alone.

NOTE: Someone should do the same thing for the Darwin stdenv if
they want to cross-compile things on Darwin.

DavidEGrayson committed:
Fixed how a bunch of packages are built when crossSystem is present.

gcc5:
- Make it so crossSystem and libcCross do not affect the native compiler.
- NOTE: We should do this for the other gcc versions too.

gccCrossStageStatic, gccCrossStageFinal:
- Use the real stdenv instead of using bootstrap tools.
  Bootstrap tools are a threat to the purity of the Nix system, so we
  should only use them to build the stdenv and nothing else.  Also,
  this seemed to fix a configure error I was getting with cross glibc
  for the Raspberry Pi.
- Explicitly disable langObjC and langObjCpp.

glibcCross:
- Use the real stdenv instead of using bootstrap tools.

libcCross1:
- Move the logic for picking this so it is closer to libcCrossChooser.

libiconv:
- Choose the nativeDrv and crossDrv independently so that the crossSystem
  does not affect the nativeDrv.

At this point, cross-compiling seems to work fine.  I tested it by
applying nix-build to this expression:

rec {
  myNixpkgsFunc = import ./nixpkgs;
  myNixpkgs = myNixpkgsFunc {};

  baseNixpkgsFunc = import ((import <nixpkgs> { }).fetchFromGitHub {
    owner = "NixOS";
    repo = "nixpkgs";
    rev = "d4eac0278ca188aa236b6a7f0fbb15f8ee4273d0";
    sha256 = "0mpx6bjvix7dfmgn6a2chmmicamz9v941fki0lgzfk2gw90fqg6w";
  });
  baseNixpkgs = baseNixpkgsFunc {};

  myTestsPass = (import ./nixpkgs/test/cross_system.nix).testsPass;

  gcc6Same = myNixpkgs.gcc6 == baseNixpkgs.gcc6;
  gcc5Same = myNixpkgs.gcc5 == baseNixpkgs.gcc5;
  gcc49Same = myNixpkgs.gcc49 == baseNixpkgs.gcc49;
  gcc48Same = myNixpkgs.gcc48 == baseNixpkgs.gcc48;
  gcc45Same = myNixpkgs.gcc45 == baseNixpkgs.gcc45;
  gccSame = gcc6Same && gcc5Same && gcc49Same && gcc45Same;

  testsPass = myTestsPass && gccSame;

  derivations = rec {
    recurseForDerivations = assert testsPass; true;

    rpiCrossSystem = {
      config = "armv6l-unknown-linux-gnueabi";
      bigEndian = false;
      arch = "arm";
      float = "hard";
      fpu = "vfp";
      withTLS = true;
      libc = "glibc";
      platform = myNixpkgs.platforms.raspberrypi;
      openssl.system = "linux-generic32";
      gcc = {
        arch = "armv6";
        fpu = "vfp";
        float = "softfp";
        abi = "aapcs-linux";
      };
    };
    rpiPkgs = myNixpkgsFunc { crossSystem = rpiCrossSystem; };
    rpiHello = rpiPkgs.hello.crossDrv;
  };
}

DavidEGrayson committed:
Added test/cross_system.nix to us make sure that the crossSystem
argument does not affect native derivations.

DavidEGrayson commented:
I'll also ping @viric because it sounds like he was doing cross-compiling.

bjornfor commented:
Also @dezgeg, @Ericson2314, @rasendubi

Ericson2314 commented:
I will take a look at this tomorrow but you may be interested in https://github.com/NixOS/nixpkgs/pull/15043 and https://github.com/Ericson2314/nixpkgs/commit/ee2a4842175f31aeef2c386a53ac65f7c8f7c3b7

groxxda commented on pkgs/development/compilers/gcc/5/default.nix:
```diff
@@ -547,4 +547,10 @@ stdenv.mkDerivation ({
+# want to avoid a meaningless mass rebuild.  These variables get
+# overridden via crossAttrs in an actual cross-build.
+// { crossMingw = false; crossStageStatic = true; }
+
```
So this is only required one time for the transition? If yes I think we should remove and deal with one mass-rebuild :wink:

DavidEGrayson replied:
I thought that adding this would make it simpler and easier for my pull request to be merged, and it also helped me ensure that I did not accidentally screw something up.  I used Nix assertions to [assert](https://gist.github.com/DavidEGrayson/8d61032e523b3bc8bbca0c4b281afb52) that the native GCC in my pull request and in the upstream commit it is based on are the same :-).

This is kind of a subjective decision, so I would mainly be concerned with what the people in charge of nixpkgs think.  @groxxda, do you authority to merge my pull request?

groxxda replied:
I'm not in the position to merge but nixpkgs is a community project and I consider myself part of the community just like any other contributor :wink: 
This is my subjective view:
Avoiding a single mass rebuild does not weight out the cons of "dead code" (whereas during testing it is very valid), and one mass rebuild is not a blocker for pull requests to get merged.
Since we only need this code to make the test work, it should be moved into the test file.

DavidEGrayson replied:
Are there instances in the past where we chose to do a mass rebuild so we could remove one line of dead code?

Also, does your subjective opinion change if I point out that the recipe for GCC 5 will be deleted some day when nothing depends on it any more?

groxxda replied:
This can be merged with an update that requires a mass-rebuild on it's own so the extra cost could be minimal.
This line needs to be duplicated into every gcc expression that should be asserted inside the tests, right? Imo that's an even stronger argument for moving it inside the tests file and make it generic.
For me it's as simple as: If you only need it for the tests, put it inside the tests file.

DavidEGrayson replied:
It's only needed to make my [outside-of-tree tests](https://gist.github.com/DavidEGrayson/8d61032e523b3bc8bbca0c4b281afb52) pass and to avoid a mass rebuild.  The tests in the tree would not be able to detect that a mass rebuild is happening because they only look at the current commit of nixpkgs.  The tests in the tree just make sure that a few native derivations don't get changed when crossSystem is specified.

rasendubi replied:
I'm also in favor of mass rebuild here. Mass-rebuild patches are merged into staging, so they don't disturb normal much.

Also if we consider the fact that we'll remove that dead code in the future, that will cause mass rebuild anyway. Don't see any reason to wait.

groxxda commented on pkgs/stdenv/linux/default.nix:
```diff
@@ -284,22 +284,13 @@ rec {
+      "findutils" "gawk" "gcc" "glibc" "gnugrep" "gnumake" "gnused"
+      "gnutar" "gnupatch" "gzip" "patchelf" "paxctl" "pcre" "xz"
+      "zlib"
+    ] stage4.pkgs;
```
Why do we need strings here?

DavidEGrayson replied:
Actually, I should be able to replace the first two attributes of lib.overrideNativeDrvs with a single set is just `{ inherit (stage4.pkgs) attr acl ... zlib; }`.  Do people like that better?

Ericson2314 replied:
Why is `allowedRequisites` being deleted? IIRC it is some thing to lint the dependencies the derivation has, but I could be wrong.. The commented code is probably horribly bitrotted and probably should be removed with an issue created, but strictly speaking that change his little to do with this PR?

groxxda commented on pkgs/top-level/default.nix:
```diff
@@ -57,8 +57,8 @@ let
     else config.platform or platformAuto;
 
-  topLevelArguments = {
-    inherit system bootStdenv noSysDirs config crossSystem platform lib;
```
Not sure why you changed the name, but it looks like you add lib later in all cases, or am I missing something?

DavidEGrayson replied:
Sorry for not commenting on that in the commit message.  I changed `topLevelArguments` to `topLevelArgs` to be more consistent with other variables in the same file that are named `*Args`.  

I took `lib` out of that set because `lib` is not actually a top-level argument; it's not argument to `top-level/default.nix`.  Also, I needed to omit lib from that set on this line:

    allPackages = args: import ./. (topLevelArgs // args);

I hope that makes sense.

DavidEGrayson commented:
@Ericson2314 I haven't looked at your [pull request](https://github.com/NixOS/nixpkgs/pull/15043) but it's good to see that some other people care about getting cross-compiling to work.

groxxda commented:
Please don't get me wrong, I'm really :+1: to see this merged, since as far as I can tell cross compilation is broken and most of the other work got stale before it was ready to merge.
Since this PR looks minimal and get's things working, I :pray: this finally makes it into master.

shlevy commented:
@DavidEGrayson Do you know if anyone who has had cross-compilation working in the past now sees it failing? Note that it's definitely worked before, even if it doesn't currently, so it may just be something documented wrong.

Ericson2314 commented:
@DavidEGrayson did you take into account the bootstrap phases? the stdenvs themselves still call all packages, so as far as I can tell the importing in both directions occurs.

Ericson2314 commented:
@shlevy In my experience it does work, but there is pointless rebuilding currently.

DavidEGrayson commented:
@shlevy: No I don't, but I'll try to look.  Maybe I'll try to figure out how to build `pkgs/top-level/release-cross.nix` and see if that works in the current master.

DavidEGrayson commented:
@Ericson2314: Yes, the allPackages function is used by stdenv to bootstrap itself, but from the perspective of the stdenv expressions, that is now just a simple function call instead of an actual `import`, so I think it's easier to think about.

Ericson2314 commented:
@DavidEGrayson on the second part, yeah, talking about my own.

DavidEGrayson commented:
@rasendubi: This pull request moves nixpkgs closer to the point that you would want it to be, right?  So would you support merging the pull request as is?  (You could make a small pull request later where you argue for the mass-rebuild, and label it with the the [appropriate severity label](https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+is%3Apr+label%3A%221.severity%3A+mass-rebuild%22).)

Ericson2314 commented:
CC @vcunat and @Mathnerd314, who were interested in this sort of things.

Ericson2314 commented:
I'm looking at the commits in detail now.

Ericson2314 commented:
@DavidEGrayson I like the first two commits and the tests, for sure. I think the 3rd commit may be better addressed in by adding a stage to the cross stdenv like in my PR, and using `overrideNativeDeps`, but I'm still thinking about it. That part of my PR should be relatively orthogonal to the rest and able to be cherry picked.

rasendubi commented:
@DavidEGrayson I'm no way blocking the pull request. Though, I don't support it (yet), as I would like to test it/review in detail first. (Hope to do that at weekend.)

DavidEGrayson commented:
My TODO list for tonight:

- [ ] Don't use a long list of strings in pkgs/stdenv/linux/default.nix to defined stdenvOverrides.
- [ ] Restore the comment about allowedRequisites in pkgs/stdenv/linux/default.nix even though it looks junky.

DavidEGrayson committed:
pkgs/stdenv/linux/default.nix: Restore an old comment
about allowedRequisites that I did not write.

DavidEGrayson committed:
pkgs/stdenv/linux/default.nix: Express the native derivation overrides
using a set instead of a list of strings.  This is more flexible and
looks nicer.

DavidEGrayson commented:
Thanks to some info from @shlevy, I have looked at `pkgs/top-level/release-cross.nix`, and I have learned that it can generally be used to successfully cross-compile some things, and Hydra does that regularly with the [cross-trunk job set](https://hydra.nixos.org/jobset/nixpkgs/cross-trunk).

But here is a concrete example of something that fails for me on nixpkgs master.  Suppose I check out a recent commit on master, d4eac0278ca188aa236b6a7f0fbb15f8ee4273d0, and add this file in the `pkgs/top-level/` directory:

```
(import ./. {
  system = builtins.currentSystem;
  crossSystem = {
    config = "foobar";
    arch = "arm";
    libc = "uclibc";
    platform = (import ../.. {}).platforms.raspberrypi;
  };
}).git
```

This file is not supposed to be valid, but it sort of emulates what I was trying to do when I first looked into cross-compiling with Nix.  I was trying to compile for a new system that does not use glibc, and I needed to fetch some sources from git to support building the compiler from it.  If I run nix-build on this new file, the abridged result is:

```
these derivations will be built:
  /nix/store/1bnm7156fx4vz60fs5kf2yz1wv5127jp-linux-pam-1.2.1.drv
  /nix/store/1xrc64k02gb6qgqaf84684209w1ybmg6-w3m-0.5.3-2015-12-20.drv
  /nix/store/wfgdz7qnjxaqbdair7n93rv4wm537s5m-libxml2-2.9.4.drv
  /nix/store/308l502f4w12ag7rvwqr6zgnkb7m338j-libxslt-1.1.28.drv
  /nix/store/qrc61rxalpir6wdifd01gqn7207ck2vh-xmlto-0.0.28.drv
  /nix/store/97f1q4d408adiirr87cpp3hqvhc6prrk-opensp-1.5.2.drv
  /nix/store/adsgyqf7wsvyhflfkpwf31kr5vh29wd5-libiconv-1.14.drv
  /nix/store/5ynq4i4fm56hyicq86683djjxvi1jjkb-docbook2X-0.8.8.drv
  /nix/store/v0wd59xn219zyhvqr5ysgjmvvgv9is4g-pcre-8.38.drv
  /nix/store/76rcb8zdn89wl7d5n3skmzspqxvyvm36-gnugrep-2.25.drv
  /nix/store/h6jmab5xlgdz4g6cdnqflllnymg74r8l-texinfo-5.2.drv
  /nix/store/v88l8qvzcpxgchrplqqkqdbyszskpji3-openssh-7.3p1.drv
  /nix/store/kly3x044yckwi19d3v68c3r20mkrawa9-git-2.10.0.drv
building path(s) ‘/nix/store/191sag71p88pgv4widn2zrzv838lblsr-libiconv-1.14’
unpacking sources
unpacking source archive /nix/store/mgvhamvs92cvhzkq3qwg2slz9sj2vcd7-libiconv-1.14.tar.gz
........
make[3]: Leaving directory '/tmp/nix-build-libiconv-1.14.drv-0/libiconv-1.14'
gcc -DHAVE_CONFIG_H -DEXEEXT=\"\" -I. -I.. -I../lib  -I../intl -DDEPENDS_ON_LIBICONV=1 -DDEPENDS_ON_LIBINTL=1   -g -O2 -c allocator.c
gcc -DHAVE_CONFIG_H -DEXEEXT=\"\" -I. -I.. -I../lib  -I../intl -DDEPENDS_ON_LIBICONV=1 -DDEPENDS_ON_LIBINTL=1   -g -O2 -c areadlink.c
gcc -DHAVE_CONFIG_H -DEXEEXT=\"\" -I. -I.. -I../lib  -I../intl -DDEPENDS_ON_LIBICONV=1 -DDEPENDS_ON_LIBINTL=1   -g -O2 -c careadlinkat.c
gcc -DHAVE_CONFIG_H -DEXEEXT=\"\" -I. -I.. -I../lib  -I../intl -DDEPENDS_ON_LIBICONV=1 -DDEPENDS_ON_LIBINTL=1   -g -O2 -c malloca.c
gcc -DHAVE_CONFIG_H -DEXEEXT=\"\" -I. -I.. -I../lib  -I../intl -DDEPENDS_ON_LIBICONV=1 -DDEPENDS_ON_LIBINTL=1   -g -O2 -c progname.c
In file included from progname.c:26:0:
./stdio.h:1010:1: error: 'gets' undeclared here (not in a function)
 _GL_WARN_ON_USE (gets, "gets is a security hole - use fgets instead");
 ^
make[2]: *** [Makefile:914: progname.o] Error 1
make[2]: Leaving directory '/tmp/nix-build-libiconv-1.14.drv-0/libiconv-1.14/srclib'
make[1]: *** [Makefile:865: all] Error 2
make[1]: Leaving directory '/tmp/nix-build-libiconv-1.14.drv-0/libiconv-1.14/srclib'
make: *** [Makefile:35: all] Error 2
builder for ‘/nix/store/adsgyqf7wsvyhflfkpwf31kr5vh29wd5-libiconv-1.14.drv’ failed with exit code 2
cannot build derivation ‘/nix/store/kly3x044yckwi19d3v68c3r20mkrawa9-git-2.10.0.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/kly3x044yckwi19d3v68c3r20mkrawa9-git-2.10.0.drv’ failed
```

So we can see that because of some values I chose for the crossSystem, nixpkgs thinks it needs to build a native libiconv for its native git package, and it fails to build the native libiconv.  When that happened, I decided that even though my crossSystem values were not completely valid, nixpkgs was behaving badly and I needed to debug nixpkgs itself to make progress.

I hope you can now understand my motivation for this pull request.

shlevy commented:
@DavidEGrayson To be clear, technically nixpkgs is not buggy but is poorly documented and has a bad interface, right? So you're improving the interface? (have not yet had time to fully review)

DavidEGrayson commented:
I am improving the internal interface between `pkgs/top-level/default.nix` and `pkgs/top-level/stdenv.nix`, but that isn't the most important part of the pull request.  That is the first commit of the pull request.  The other commits fix buggy behavior, such as the native GCC derivation depending on `crossSystem` and other things like that.

DavidEGrayson commented:
I've discovered a Hydra build job that used to pass and would be broken by now by tests:

```
nix-build pkgs/top-level/release-cross.nix -A bootstrapTools.armv7l
error: attribute ‘crossDrv’ missing, at /home/david/nixpkgs/pkgs/stdenv/linux/make-bootstrap-tools-cross.nix:89:22
```

(That error happens on my branch but not on d4eac0278ca188aa236b6a7f0fbb15f8ee4273d0.)

I will need some time to look into that and make it pass again.  You can still merge my pull request in the mean time if you want.

DavidEGrayson committed:
libtool: Fix it so the native derivation does not depend on
crossSystem.

In general, none of the native derivations should change when
crossSystem is specified.

DavidEGrayson commented:
Hey, what do people think of the [libxml2 recipe](https://github.com/NixOS/nixpkgs/blob/d4eac0278ca188aa236b6a7f0fbb15f8ee4273d0/pkgs/development/libraries/libxml2/default.nix)?  I think that in order to make the native libxml2 derivation *not* depend on crossSystem, we'll need to make major changes to it.  The `crossAttrs` mechanism alone is not powerful enough to do what we want to do in the libxml2 recipe.  It looks like the author of that recipe wants to set `supportPython` to false when we are cross-compiling libxml2, but they did it wrong,  so whenever crossSystem is specified, the native derivation of libxml2 ends up having supportPython = false, which causes needless rebuilding (and possible breakage) of a lot of native packages.

Perhaps I need to make a function in the libxml2 recipe that takes a boolean argument indicating whether we are cross-compiling or not, and then I would make the derivation like this:

    stdenv.mkDerivation (libxml2Attrs false) // { crossAttrs = (libxml2Attrs true); }

Alternatively, we could make a new version of mkDerivation that helps support cases like this, which are probably quite common.

bjornfor commented:
Another PR (old and unmerged) improving libiconv + cross: https://github.com/NixOS/nixpkgs/pull/15021.

DavidEGrayson committed:
Added mkDerivationWithCrossArg in order to fix the libxml2 derivation
and make it not depend on the crossSystem argument.

DavidEGrayson committed:
Fix the dontStrip attribute of several more packages.

DavidEGrayson commented:
My solution to libxml2 looks like this:

```
{ stdenv, lib, fetchurl, zlib, xz, python, findXMLCatalogs, libiconv, fetchpatch
, supportPython ? null }:

stdenv.mkDerivationWithCrossArg(cross:
  let
    supportPythonReally = if supportPython == null
                          then cross == null
                          else supportPython;
```

Here is my current TODO list:

- [ ] Add mkDerivationWithCrossArg.
- [ ] Fix some more packages whose native derivations depended on the `crossSystem` argument.
- [ ] Fix all the Hydro jobs I broke in release-cross.nix.
- [ ] Integrate my tests in `test/cross_system.nix` with `release-cross.nix` so they automatically get run by Hydra. Who besides me would like to be emailed when those tests fail?

After all that, I think this pull request would be ready to merge, unless you guys can think of some other important thing that needs to be verified or tested.

groxxda commented:
atm this is limited to gcc5, is it possible to make it work for any gcc?

DavidEGrayson commented:
@groxxda Are you talking about cross-compiling using a version of GCC other than 5?  I think that would probably work.  You would have to change the logic in `all-packages.nix` for the `gccCrossStageStatic` and `gccCrossStageFinal`.  I don't think my pull request makes that easier or harder.  Previously we just hard-coded the policy of using the same gcc for the native system and the cross system.  Now, in this pull request, I am hard-coding the policy of using gcc 5 for the cross system, but we'll probably update that whenever we update the gcc used for the native system, which is also hard-coded to be gcc5.

I imagine you might be able to make it easier to override cross gcc by having an attribute like this `all-packages.nix` that you override, but I am not sure:

    crossGccExpression = ../development/compilers/gcc/5;

DavidEGrayson committed:
overrideNativeDrv: Provide an override function so that the expression
for coreutilsMinimal in make-bootstrap-tools-cross.nix works again.

DavidEGrayson committed:
make-bootstrap-tools-cross.nix: Fix it to use pkgs.gcc.crossDrv.cc
instead of pkgs.gcc.cc.crossDrv, which does not exist anymore.

DavidEGrayson committed:
GCC 5: Use --with-sysroot instead of --with-headers for cross-compilers.

The GCC 5 documentation says that --with-headers is obsolete and
deprecated in favor of --with-sysroot.  Also, this seems to fix a
problem I was having where the configure script could not find glibc's
limit.h, so it did not add the administrivia at the top and bottom of
include-fixed/limits.h which tell it to #include_next the real
limits.h, which resulted in limits.h defining the wrong value for
MB_LEN_MAX, which caused an error while cross-building coreutils.

DavidEGrayson commented:
So actually the `crossMingw` and `crossStageStatic` attributes outside of `crossAttrs` in the GCC 5 expression are *not* dead code: they are needed when you are building a cross-compiler, and when you are building a cross-compiler, you not actually cross-compiling yet, so `crossAttrs` and `crossDrv` are not in play.

I am still working on fixing up this pull request as described above in my TODO list.

DavidEGrayson commented:
Hey, the `--with-headers` argument to GCC's configure script, which [we are using for cross-compilation](https://github.com/NixOS/nixpkgs/blob/d4eac0278ca188aa236b6a7f0fbb15f8ee4273d0/pkgs/development/compilers/gcc/5/default.nix#L170) is deprecated according to the in-tree documentation of GCC 5.4.0 and 6.2.0:

https://github.com/gcc-mirror/gcc/blob/gcc-5_4_0-release/gcc/doc/install.texi#L2020-L2022
https://github.com/gcc-mirror/gcc/blob/gcc-6_2_0-release/gcc/doc/install.texi#L2069-L2071

The documentation says you are supposed to use `--with-sysroot` instead.  Assuming that `--with-sysroot` works, are there any objections to changing the recipe to use it in this pull request, even if it is not strictly necessary?  This would only affect cross-compilation.  One of the benefits is that the headers of the cross libc would not get copied into the GCC output, so we would save some space.

shlevy commented:
@DavidEGrayson :+1: on switching configure option

DavidEGrayson committed:
GCC 5 package: temporary fix to get cross-compiling working

DavidEGrayson committed:
GCC 5 recipe: Major changes to make it closer to the upstream version and more likely to be correct.

DavidEGrayson committed:
gcc 5: Change the cross MinGW static-stage options back because GCC was having trouble finding the headers.

DavidEGrayson commented:
My Raspberry Pi bootstrap tools are compiling for "armv5t" instead of "arm" for some reason, even though the compiler is named armv6l-unknown-linux-gnueabi-gcc.  Debugging the cross GCC is harder than it needs to be because of this issue which I just reported: https://github.com/NixOS/nixpkgs/issues/19151

bjornfor commented:
Any news?

DavidEGrayson commented:
Not much.  But I started building a [drvdiff utility](https://github.com/DavidEGrayson/drvdiff) to help me figure out why my Raspberry Pi cross-compiler would behave so differently than the upstream nixpkgs one (producing output for the wrong ARM architecture).

zsmith928 commented:
@DavidEGrayson @bjornfor My company is offering sponsored on-demand ARM64 hardware if you're interested in testing your build on AARCH64.  WE can support standard CI tools to deploy the hosts automatically, etc, so DM me if you're in need of access for NIX.

shlevy commented:
Status on this? Please ping me when you think this is ready

DavidEGrayson commented:
The status of this pull request is still the same as last time: I am (slowly) working on this [drvdiff utility](https://github.com/DavidEGrayson/drvdiff) so I can efficient debug why my cross toolchain would compile for the wrong ARM architecture when I try to compile "hello" for the Raspberry Pi.  Pull request #19940 has been merged recently, so this pull request will need some work to avoid conflicts when we finally get it working.

I think the right order to do things is to get drvdiff working, figure out why this pull request did not work, fix it, and then work on rebasing the whole thing onto the latest nixpkgs master and resolving conflicts.

I did some more work for this pull request back in September but I had not pushed those commits to the pull request branch because I was intending to clean them up.  Well, to make it easier for other people who want to help out, I have now pushed those commits to the pull request branch (updating that branch from 873da5a to 9cedf1b).

DavidEGrayson commented:
Here are my **[notes]** and **[test NIX expression]** for the GCC ARM architecture problem in case anyone wants to investigate.

[notes]: https://gist.github.com/DavidEGrayson/a94c4df67632966553ab72e9f9c7b134
[test NIX expression]: https://gist.github.com/DavidEGrayson/8d61032e523b3bc8bbca0c4b281afb52

dezgeg commented:
In `pkgs/development/compilers/gcc/5/default.nix`:

````
130     crossConfigureFlags = let
131         gccArch = stdenv.cross.gcc.arch or null;
132         gccCpu = stdenv.cross.gcc.cpu or null;
133         gccAbi = stdenv.cross.gcc.abi or null;
134         gccFpu = stdenv.cross.gcc.fpu or null;
135         gccFloat = stdenv.cross.gcc.float or null;
136         gccMode = stdenv.cross.gcc.mode or null;
137         withArch = if gccArch != null then " --with-arch=${gccArch}" else "";
138         withCpu = if gccCpu != null then " --with-cpu=${gccCpu}" else "";
139         withAbi = if gccAbi != null then " --with-abi=${gccAbi}" else "";
140         withFpu = if gccFpu != null then " --with-fpu=${gccFpu}" else "";
141         withFloat = if gccFloat != null then " --with-float=${gccFloat}" else "";
142         withMode = if gccMode != null then " --with-mode=${gccMode}" else "";
143       in
144         "--target=${cross.config}" +
145         withArch +
146         withCpu +
147         withAbi +
148         withFpu +
149         withFloat +
150         withMode +
````

After your changes `stdenv.cross` is `null` here (in the case of `gcc-5.4.0-armv6l-unknown-linux-gnueabi-stage-final.drv`) so all the `withFoo` flags are empty. The `cross` variable seems to contain the correct things though. So that needs either `s/stdenv.cross/cross/` or figuring out why `stdenv.cross` is not set if the difference was unintentional.

DavidEGrayson committed:
GCC 5: In crossConfigureFlags, use cross instead of stdenv.cross
because stdenv.cross is nullwhen we build the cross compiler.

DavidEGrayson commented:
Thanks for figuring that out, @dezgeg!  After I fixed that (commit 9fde656117923b3e31df26fbfadce9611bb36945 of this PR), I was able to successfully cross-compile a "Hello world" program for the Raspberry Pi and run it.

I fixed it by changing `stdenv.cross` to `cross` in the definition of `crossConfigureFlags` for GCC 5.  The reason `stdenv.cross` is null is because this PR removes some tricky circular references from nixpkgs: we use a normal stdenv without any cross-compilation (stdenv.cross is null) to build our cross compiler.  So when we are building the cross-compiler, `stdenv.cross` will be null.  Once the cross-compiler is built, then we create a stdenv with a `cross` attribute, as before.

Ericson2314 commented:
I rebased this (untested though) as github.com/Ericson2314/nixpkgs/tree/deg-cross

DavidEGrayson commented:
My rebase of this pull request onto a modern nixpkgs master commit is almost finished.  But while testing it, I noticed that six days ago there was a change that broke 31 of the cross-compilation jobs:

https://hydra.nixos.org/eval/1314290

DavidEGrayson commented:
I'm closing this because it's stale.  Anyone interested in the changes from this pull request should discuss them in the new version of the pull request (v2) here: https://github.com/NixOS/nixpkgs/pull/21388

dezgeg commented:
This should fix the coreutils cross building breakage: https://github.com/NixOS/nixpkgs/commit/30074ddfeac5e165eea32b33829bc2bf4185aace
